### DISCUSSION TITLE
Unknow property 'input'

### DISCUSSION THREAD

**Discussion by peter-budo** [detekt/detekt#3917] [General]
Recently I had to do some upgrades to our Android project setup as we wanted to remove `jCenter` and that kicked bit of fuss with detekt setup. We use external file outside of Android plugin as we have multiple modules in which we want to run detekt checks. This file after some edits from me looks like this
```
buildscript {
    repositories {
        gradlePluginPortal()
    }
    dependencies {
        classpath "io.gitlab.arturbosch.detekt:detekt-gradle-plugin:1.17.1"
    }
}

apply plugin : io.gitlab.arturbosch.detekt.DetektPlugin


repositories {
    mavenCentral()
}

task detektAll(type: io.gitlab.arturbosch.detekt.Detekt) {
    config.from(files("${project.rootDir}/code_quality/detekt-config.yml"))
    baseline = file("${project.rootDir}/code_quality/detekt-baseline.xml")
    exclude ".*test.*"
    exclude ".*/resources/.*"
    exclude "*/build/.*"
    input = files("$projectDir/src/main/java")
    reports {
        xml.enabled = false
        html {
            enabled = true
            destination = file("build/reports/detekt/detekt.html")
        }
    }
}
```

However when I try to assemble build I do run in following message
```
* What went wrong:
A problem occurred evaluating script.
> A problem occurred evaluating script.
   > Could not set unknown property 'input' for task ':app:detektAll' of type io.gitlab.arturbosch.detekt.Detekt.

```

Can anybody please explain me what I am doing wrong?

**Comment by marschwar** [comment]
Please have a look at https://detekt.github.io/detekt/howto-buildupondefaultconfig.html (last paragraph). I think `source` is the property you are looking for.
 
The Detekt task does not have an `input` property but rather inherits the source property from `org.gradle.api.tasks.SourceTask`.

**Reply by peter-budo** [reply]
@marschwar according to this config https://detekt.github.io/detekt/gradle.html#kotlin-dsl-4 it does 
```
// The directories where detekt looks for source files. 
    // Defaults to `files("src/main/java", "src/test/java", "src/main/kotlin", "src/test/kotlin")`.
    input = files("src/main/java", "src/main/kotlin")  
```
Hence why asking

**Reply by marschwar** [reply]
The example you are referring to is using the `detekt` extension which defines the `input` property and maps it to the source property of the task itself. I do admit that this is not very obvious and we should consider renaming the the input property to `source` for the next major release.

**Reply by peter-budo** [reply]
Hmmm after that change I'm seeing 
```
* What went wrong:
A problem occurred configuring project ':app'.
> Failed to notify project evaluation listener.
   > com/android/build/gradle/BaseExtension
```

**Reply by peter-budo** [reply]
Looks to be same issue as already reported here https://github.com/detekt/detekt/issues/3490

**Reply by marschwar** [reply]
[comment deleted]

I am sorry, my comment was completely unrelated.

**Reply by marschwar** [reply]
With the information provided in #3490 have you been able to get it to work?

**Comment by cortinico** [comment]
> Can anybody please explain me what I am doing wrong?

Just out of curiosity, have you checked that your versions are aligned with the one listed here:
https://detekt.github.io/detekt/compatibility.html

If not, can you please post any discrepancy?

**Reply by peter-budo** [reply]
My apology, I was too busy in last 2 days with app releases and sprint planning. I will try to check it out today

**Reply by peter-budo** [reply]
@cortinico here are project tool versions
Detekt - 1.17.1 (I also tried 1.16.0 but same error as 1.17.1)
Gradle - 6.8.3
Kotlin - 1.4.20
AGP - 4.2.1
Java Target - 1.8
JDK - 1.8 that is packaged together with Android Studio

**Reply by cortinico** [reply]
The issue you're facing is the same as https://github.com/detekt/detekt/issues/3490 

Are you using an external `.gradle` file?

**Reply by peter-budo** [reply]
Yes, we do keep code quality tools on external files for better clarity.
Do you think that fix can make it in next release of 1.18.0?

**Reply by cortinico** [reply]
> Yes, we do keep code quality tools on external files for better clarity.

External gradle files don't play well with AGP and Detekt. You should consider moving to precompiled script plugins (inside the `buildSrc` project).

> Do you think that fix can make it in next release of 1.18.0?

You could use https://github.com/detekt/detekt/pull/3511 and opt-out from Detekt Android. This means that you won't have the `detektMain`, `detektTest` tasks anymore, but just a plain `detekt` task that will run without type resolution. If that is ok with you, that's probably the better approach.

A more definitive solution for working with external build files still has to be developed.

**Reply by peter-budo** [reply]
> External gradle files don't play well with AGP and Detekt. You should consider moving to precompiled script plugins (inside the buildSrc project).

I tried, but that was giving me different headache :(

**Reply by cortinico** [reply]
> I tried, but that was giving me different headache :(

I can understand but it's definitely worth to look into, especially if you plan to use Gradle Kotlin DSL. We also use them to configure detekt to run on our own codebase: https://github.com/detekt/detekt/blob/main/buildSrc/src/main/kotlin/detekt.gradle.kts
